### PR TITLE
Issue 489: Added support for custom storage option in long term storage

### LINKED ISSUE #489: Cannot configure AWS S3 Tier2 storage in Helm chart
I'd like to configure Pravega Tier2 storage to use AWS S3. My understanding from the [long term storage docs][1] is that any S3 compatible backend is supported. However, I can't seem to get this working in my Kubernetes cluster.

I'm using the following secret:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: aws-credentials
type: Opaque
stringData:
  ACCESS_KEY_ID: [AWS_ACCESS_KEY_ID]
  SECRET_KEY: [AWS_SECRET_ACCESS_KEY]
```

And the relevant Helm values for the Pravega cluster configuration:
```yaml
storage:
  longtermStorage:
    type: ecs
    ecs:
      configUri: https://s3.eu-west-2.amazonaws.com
      bucket: "my-aws-s3-bucket"
      prefix: "my-prefix"
      credentials: aws-credentials
```

Using the toy NFS example I can write to and read from a stream. However, with the above configuration the deployment of Pravega fails. Has anyone had any success with AWS S3 as a Tier2 storage?

  [1]: https://github.com/pravega/pravega-operator/blob/master/doc/longtermstorage.md#use-dell-emc-ecs-as-longtermstorage

### PR DESCRIPTION
**Pull request by anishakj**
Signed-off-by: anishakj <anisha.kj@dell.com>

### Change log description

Adding a new field for custom storage in longtermstoragespec
### Purpose of the change

 Fixes #489

### What the code does

Currently, operator supports HDFS, ECS and filesystem as longtermstorage options, added support for custom storage so that users can provide required options and env variables based on the storage type selected.

We can also pass env varaibles via `segmentStoreEnvVars`. But  in case,if we do not set `longtermstorage` type, it can cause conflicts

### How to verify it

Example of helm charts with `s3` as custom storage

```
storage:
  longtermStorage:
    type: custom
    custom:
      options:
        pravegaservice.storage.layout: "CHUNKED_STORAGE"
        pravegaservice.storage.impl.name: "S3"
        s3.bucket: bucketname
        s3.prefix: "10-11-1"
        s3.connect.config.uri.override: "false"
        s3.connect.config.uri: uri
        s3.connect.config.access.key: keyname
        s3.connect.config.secret.key: keyvalue
      env:
        TIER2_STORAGE: "S3"
        AWS_ACCESS_KEY_ID: "key"
        AWS_SECRET_ACCESS_KEY: "secret"
```

Verified that able to install pravega with `S3` as custom storage, also ran system tests

### CODE REVIEW

**Review by nishant-yt — COMMENTED** [review]
Requested changes

File: pkg/apis/pravega/v1beta1/pravega.go [hunk 1]
```diff
@@ -511,3 +514,8 @@ type HDFSSpec struct {
 	// +optional
 	ReplicationFactor int32 `json:"replicationFactor"`
 }
+
+type CustomSpec struct {
+	Options map[string]string `json:"options"`
```
[the remark this answers is not in the document]
Comment (jkhalack) [reply pkg/apis/pravega/v1beta1/pravega.go:519]: @anishakj @nishant-yt Is there a use case when **only one** of `options`,`env` blocks would have to be supplied?

**Comment by anishakj** [comment]
> > > @anishakj Could you add some (commented out) example on how this issue might be used to the helm chart (https://github.com/pravega/pravega-operator/blob/master/charts/pravega/values.yaml#L107), please?
> > 
> > 
> > have to create separate PR for that in Pravega/charts repo. Charts repo here is deprecated
> 
> Would you at least add it to the PR description?

Values need to be used in helm chart are added in PR description.

**Review by jkhalack — APPROVED** [review]

**Review by jkhalack — CHANGES_REQUESTED** [review]

**Review by jkhalack — APPROVED** [review]
LGTM

**Review by SrishT — CHANGES_REQUESTED** [review]

**Review by SrishT — APPROVED** [review]
LGTM

**Review by nishant-yt — APPROVED** [review]
LGTM

**Pull request merged**